decrepit/xts/xts.c

```diff
@@ -163,7 +164,28 @@ static int aes_xts_init_key(EVP_CIPHER_CTX *ctx, const uint8_t *key,
   }
 
   if (key) {
+    // Verify that the two keys are different.
+    //
+    // This addresses the vulnerability described in Rogaway's
+    // September 2004 paper:
+    //
+    //      "Efficient Instantiations of Tweakable Blockciphers and
+    //       Refinements to Modes OCB and PMAC".
+    //      (http://web.cs.ucdavis.edu/~rogaway/papers/offsets.pdf)
+    //
+    // FIPS 140-2 IG A.9 XTS-AES Key Generation Requirements states
+    // that:
+    //      "The check for Key_1 != Key_2 shall be done at any place
+    //       BEFORE using the keys in the XTS-AES algorithm to process
+    //       data with them."
+    //
     // key_len is two AES keys
+
+    if (OPENSSL_memcmp(key, key + ctx->key_len / 2, ctx->key_len / 2) == 0) {
```

Comment:
Compile should just replace this with a right-shift. Should we do it for it?

If we are only supporting a fixed key length - maybe this computation could even just be a constant?

Reply:
I thought to keep the same values used in the rest of the code below.

It seems to me that both the C and assembly implementations can support a 128-bit key size if we define an EVP_CIPHER for it, and it's more consistent with the rest of the code to not put constants than around the block size and input size not being less than one block (16 bytes).